sdk/typescript/src/transactions/ObjectCache.ts

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { bcs } from '../bcs/index.js';
+import type { SuiClient } from '../client/client.js';
+import type { ExecuteTransactionBlockParams } from '../client/index.js';
+import type { Signer } from '../cryptography/keypair.js';
+import { normalizeSuiAddress } from '../utils/sui-types.js';
+import type { OpenMoveTypeSignature } from './blockData/v2.js';
+import type { TransactionBlock } from './TransactionBlock.js';
+import type { TransactionBlockDataResolverPlugin } from './TransactionBlockDataResolver.js';
+
+export interface AsyncCache<T> {
+	get(key: string): Promise<T | null>;
+	set(key: string, value: T): Promise<void>;
+	delete(key: string): Promise<void>;
+	clear(): Promise<void>;
+}
+
+interface ObjectCacheEntry {
+	objectId: string;
+	version: string;
+	digest: string;
+	owner: string | null;
+	initialSharedVersion: string | null;
+}
+
+interface MoveFunctionEntry {
+	package: string;
+	module: string;
+	function: string;
+	parameters: OpenMoveTypeSignature[];
+}
+
+class InMemoryCache<T> implements AsyncCache<T> {
+	#cache = new Map<string, T>();
+
+	async get(key: string): Promise<T | null> {
+		return this.#cache.get(key) ?? null;
+	}
+
+	async set(key: string, value: T): Promise<void> {
+		this.#cache.set(key, value);
+	}
+
+	async delete(key: string): Promise<void> {
+		this.#cache.delete(key);
+	}
+
+	async clear(): Promise<void> {
+		this.#cache.clear();
+	}
+
+	static createCache<T>(): AsyncCache<T> {
+		return new InMemoryCache<T>();
+	}
+}
+
+interface ObjectCacheOptions {
+	createCache?: <T>(name: string) => AsyncCache<T>;
+}
+
+export class ObjectCache implements TransactionBlockDataResolverPlugin {
+	#createCache: <T>(name: string) => AsyncCache<T>;
+	#objects: AsyncCache<ObjectCacheEntry>;
+	#functions: AsyncCache<MoveFunctionEntry>;
+
+	constructor({ createCache = InMemoryCache.createCache }: ObjectCacheOptions = {}) {
+		this.#createCache = createCache;
+		this.#objects = this.#createCache('objects');
+		this.#functions = this.#createCache('functions');
+	}
+
+	getObjects: NonNullable<TransactionBlockDataResolverPlugin['getObjects']> = async (ids, next) => {
+		const results = new Map<string, ObjectCacheEntry>();
+
+		await Promise.all(
+			ids.map(async (id) => {
+				const cachedObject = await this.#objects.get(id);
```

Comment:
We could also make the cache take a list of keys, because some cache layers might be able to do that significantly faster

Reply:
I was wondering about that, and trying to figure out if making the cache implementation more efficient is worth making it harder to implement, but I'm happy to add that.

I could make an abtract cache class that implements a multi-get method that just invokes the get method for each one, which you can optionally override

Reply:
So for some platforms like web, this won't really matter because the underlying APIs are sync, but I think for some execution environments like backends, the performance difference could be pretty substantial.